File: internal/repository/interfaces.go
package repository

import (
	"context"
	"strings"

	"github.com/ken5scal/obsflow/internal/model"
)

type StateReader interface {
	GetCheckpoint(ctx context.Context, sourceID string) (model.Checkpoint, error)
	SeenSourceItem(ctx context.Context, sourceID string, itemKey string) (bool, error)
	SeenContentHash(ctx context.Context, sourceID string, contentHash string) (bool, error)
	LastJobRun(ctx context.Context, jobID string) (model.JobRun, error)
}

type StateWriter interface {
	PutCheckpoint(ctx context.Context, cp model.Checkpoint) error
	MarkSourceItemSeen(ctx context.Context, sourceID string, itemKey string, contentHash string) error
	SaveJobRun(ctx context.Context, run model.JobRun) error
}

type StateTx interface {
	StateReader
	StateWriter
}

type StateRepository interface {
	StateTx
	InTx(ctx context.Context, fn func(tx StateTx) error) error
	Close() error
}

type VaultRepository interface {
	UpsertRecord(ctx context.Context, record model.Record) (string, error)
	ReadRecord(ctx context.Context, path string) ([]byte, error)
	UpdateAISummary(ctx context.Context, path string, summary string) error
}

type RSSClient interface {
	Fetch(ctx context.Context, sourceID string, feedURL string) ([]model.SourceItem, error)
	Parse(ctx context.Context, sourceID string, data []byte) ([]model.SourceItem, error)
}

type XClient interface {
	Search(ctx context.Context, sourceID string, query string) ([]model.SourceItem, error)
	Lists(ctx context.Context, sourceID string, listID string) ([]model.SourceItem, error)
	Bookmarks(ctx context.Context, sourceID string) ([]model.SourceItem, error)
}

type AIClient interface {
	Summarize(ctx context.Context, record model.Record) (string, error)
	BuildDigest(ctx context.Context, cadence string, records []model.Record) (string, error)
}

type AlertClient interface {
	SendError(ctx context.Context, alert Alert) error
}

type Alert struct {
	TickRunID string
	Failures  []AlertFailure
}

type AlertFailure struct {
	TargetID     string
	ErrorSummary string
	RetryHint    string
}

func (a Alert) Summary() string {
	parts := make([]string, 0, len(a.Failures))
	for _, failure := range a.Failures {
		parts = append(parts, failure.TargetID+": "+failure.ErrorSummary)
	}
	return strings.Join(parts, "; ")
}

File: src/adapters/interfaces.ts
import type {
  BaseConfig,
  Checkpoint,
  JobRun,
  SourceItem,
  VaultRecord,
  AiSummaryResult,
  FailureReport,
} from "../types.js";

export interface StateTx {
  getCheckpoint(sourceId: string): Checkpoint | null;
  putCheckpoint(cp: Checkpoint): void;
  seenSourceItem(sourceId: string, itemKey: string): boolean;
  markSourceItemSeen(sourceId: string, itemKey: string, contentHash: string): void;
  seenContentHash(sourceId: string, contentHash: string): boolean;
}

export interface StateRepository {
  getCheckpoint(sourceId: string): Promise<Checkpoint | null>;
  putCheckpoint(cp: Checkpoint): Promise<void>;
  seenSourceItem(sourceId: string, itemKey: string): Promise<boolean>;
  markSourceItemSeen(
    sourceId: string,
    itemKey: string,
    contentHash: string,
  ): Promise<void>;
  seenContentHash(sourceId: string, contentHash: string): Promise<boolean>;
  lastJobRun(jobId: string): Promise<JobRun | null>;
  saveJobRun(run: JobRun): Promise<void>;
  /** Synchronous transaction body; do not await inside `fn`. */
  inTx<T>(fn: (tx: StateTx) => T): Promise<T>;
  tryAcquireTickLock(tickRunId: string, staleMs: number): boolean;
  releaseTickLock(tickRunId: string): void;
  close(): Promise<void>;
}

export interface RssAdapter {
  collect(sourceId: string): Promise<SourceItem[]>;
}

export interface XCollectorAdapter {
  collectSearch(sourceId: string, query: string): Promise<SourceItem[]>;
  collectList(sourceId: string, listId: string): Promise<SourceItem[]>;
  collectBookmarks(sourceId: string): Promise<SourceItem[]>;
}

export interface VaultAdapter {
  upsertRecord(record: VaultRecord, noteRelPath: string): Promise<void>;
  upsertBase(base: BaseConfig): Promise<void>;
  updateAiSummary(
    noteRelPath: string,
    aiSummaryMarkdown: string,
    patch: Partial<Pick<VaultRecord, "summary" | "status" | "updated_at" | "tags" | "category">>,
  ): Promise<void>;
  listNotePathsUnder(prefix: string): Promise<string[]>;
  readRecord(noteRelPath: string): Promise<VaultRecord | null>;
}

export interface AiAdapter {
  summarize(record: VaultRecord): Promise<AiSummaryResult>;
}

export interface AlertAdapter {
  notifyFailure(report: FailureReport): Promise<void>;
  /** Flush aggregated duplicates for this tick. */
  flush?: () => Promise<void>;
}
